tests/unit/models/transactions/test_did_set.py

```diff
@@ -62,3 +62,25 @@ def test_empty(self):
                 error.exception.args[0],
                 "{'did_set': 'Must have one of `did_document`, `data`, and `uri`.'}",
             )
+
+    def test_all_empty_fields(self):
+        with self.assertRaises(XRPLModelException):
+            DIDSet(account=_ACCOUNT, data="", did_document="", uri="")
+
+    def test_delete_multiple_fields(self):
+        # create a valid DID object
+        tx = DIDSet(
```

Comment:
All of these transactions should be in separate tests.

Reply:
there are only two transactions -- create a `DID` object, and then remove the `data` field from the said object.

I cannot decouple them into separate unit tests, I need a `DID` ledger object to demonstrate the removal of the `data` field.

Reply:
It's a unit test, the ledger object is never created in the first place.

Reply:
okay, that's true.

should I remove this unit test altogether? it appears that rippled would have tested this piece of functionality.
I don't see other ways of testing this through a unit test

Reply:
Testing that one empty field works is still important.

Reply:
okay. you mean like creating a DID object with one empty field?

```
def test_single_empty_field(self):
        # create a valid DID object
        tx = DIDSet(
            account=_ACCOUNT,
            did_document=_VALID_FIELD,
            uri="", # URI field is empty
            data=_VALID_FIELD,
        )
        self.assertTrue(tx.is_valid())
```

Reply:
Yes. Or without including the `did_document` and `data` fields.

Reply:
I gave this input to one of the integration tests:
```
tx = DIDSet(account=WALLET.address, uri="")
```

This fails with an error -- we cannot input an empty URI field, while other fields are also non-existent.
```
======================================================================
FAIL: test_all_fields (tests.integration.transactions.test_did_set.TestDIDSet.test_all_fields) (version='async', client='json')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ckeshavabs/xrpl-py/tests/integration/it_utils.py", line 290, in modified_test
    asyncio.run(
  File "/Users/ckeshavabs/.pyenv/versions/3.11.6/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/Users/ckeshavabs/.pyenv/versions/3.11.6/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ckeshavabs/.pyenv/versions/3.11.6/lib/python3.11/asyncio/base_events.py", line 653, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/Users/ckeshavabs/xrpl-py/tests/integration/it_utils.py", line 281, in _run_async_test
    raise e
  File "/Users/ckeshavabs/xrpl-py/tests/integration/it_utils.py", line 277, in _run_async_test
    await test_function(self, client)
  File "/Users/ckeshavabs/xrpl-py/tests/integration/transactions/test_did_set.py", line 22, in test_all_fields
    self.assertEqual(response.result["engine_result"], "tesSUCCESS")
AssertionError: 'tecEMPTY_DID' != 'tesSUCCESS'
- tecEMPTY_DID
+ tesSUCCESS
```

Reply:
You're running it as an integration test, not a unit test. I don't need this to be tested as an integration test, I know that's going to work. That's what rippled unit tests are for. I just want to make sure that xrpl-py accepts it as a valid transaction, without knowing anything about the ledger state.

Reply:
okay. I've updated this in 10f7f38

I am slightly opposed to this type of tests. I'd prefer to have unit tests that are in consonance with the integration tests too, that reduces ambiguity for future developers

Reply:
> that reduces ambiguity for future developers

What do you mean by this?

The way I think about it is that unit tests are checking xrpl-py Python code and integration tests are ensuring that xrpl-py Python code matches rippled code. So for checking xrpl-py error code, you only want to write unit tests.

Reply:
that is true.

When looking at the unit tests, I'm led to believe that a DID object can be created with empty fields. But that is not the case with a real rippled instance.

I feel this contradiction could have been avoided, but it's not a strong objection

Reply:
I would assume anyone advanced enough to be looking at unit tests to understand the system is probably well-versed enough in the XRPL that they're not going to fall into that kind of trap - or will at least know that unit tests aren't being submitted to the ledger.

Reply:
fair point